samza-hdfs/src/main/java/org/apache/samza/system/hdfs/descriptors/HdfsSystemDescriptor.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.system.hdfs.descriptors;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.samza.system.descriptors.SystemDescriptor;
+import org.apache.samza.system.hdfs.HdfsConfig;
+import org.apache.samza.system.hdfs.HdfsSystemFactory;
+
+
+/**
+ * A {@link HdfsSystemDescriptor} can be used for specifying Samza and HDFS-specific properties of a HDFS
+ * input/output system. It can also be used for obtaining {@link HdfsInputDescriptor}s and
+ * {@link HdfsOutputDescriptor}s, which can be used for specifying Samza and system-specific properties of
+ * HDFS input/output streams.
+ * <p>
+ * System properties provided in configuration override corresponding properties specified using a descriptor.
+ */
+public class HdfsSystemDescriptor extends SystemDescriptor<HdfsSystemDescriptor> {
+  private static final String FACTORY_CLASS_NAME = HdfsSystemFactory.class.getName();
+
+  private Optional<String> datePathFormat = Optional.empty();
+  private Optional<String> outputBaseDir = Optional.empty();
+  private Optional<Long> writeBatchSizeBytes = Optional.empty();
+  private Optional<Long> writeBatchSizeRecords = Optional.empty();
+  private Optional<String> writeCompressionType = Optional.empty();
+  private Optional<String> writerClass = Optional.empty();
+
+  private Optional<Long> consumerBufferCapacity = Optional.empty();
```

Comment:
I would prefer if we add all the consumer/producer configs in the the input and output descriptor respectively instead of the system descriptor.

Reply:
This is another unfortunate unresolved issue in HDFS. Where all the configs are per system not per stream. Once we move to per stream config then it definitely makes sense to place them respectively in input and output descriptor.